ui/src/app/edge/history/soc/soc.component.ts

```diff
@@ -88,19 +87,158 @@ export class SocComponent extends AbstractHistoryChart implements OnInit, OnChan
+            });
+          }
+
+          if ('_sum/EssActivePower' in result.data) {
+            /*
+             * Storage Discharge
+             */
+            let effectivePower;
+
+            if ('_sum/ProductionDcActualPower' in result.data && result.data['_sum/ProductionDcActualPower'].length > 0) {
+              effectivePower = result.data['_sum/ProductionDcActualPower'].map((value, index) => {
+                return Utils.subtractSafely(result.data['_sum/EssActivePower'][index], value);
+              });
+            } else {
+              effectivePower = result.data['_sum/EssActivePower'];
+            }
+
+            dischargeData = effectivePower.map(value => {
+              if (value == null) {
+                return null
+              } else if (value > 0) {
+                return value;
+              } else {
+                return 0;
+              }
+            });
+
+
+          }
+
+          if ('_sum/GridActivePower' in result.data) {
+            /*
+             * Buy From Grid
+             */
+            buyFromGridData = result.data['_sum/GridActivePower'].map(value => {
+              if (value == null) {
+                return null
+              } else if (value > 0) {
+                return value;
+              } else {
+                return 0;
+              }
+            })
+
+            /*
+             * Sell To Grid
+             */
+            sellToGridData = result.data['_sum/GridActivePower'].map(value => {
+              if (value == null) {
+                return null
+              } else if (value < 0) {
+                return value * -1; // invert value
+              } else {
+                return 0;
+              }
+            });
+          };
+
+          if ('_sum/ProductionActivePower' in result.data) {
+            /*
+             * Production
+             */
+            productionData = result.data['_sum/ProductionActivePower'].map(value => {
+              if (value == null) {
+                return null
+              } else {
+                return value;
+              }
+            });
+          }
+
           if ('_sum/EssSoc' in result.data) {
             /*
-            * State-of-charge
-            */
+             * State-of-charge
+             */
             datasets.push({
               label: this.translate.instant('General.Soc'),
               data: result.data['_sum/EssSoc'],
               hidden: false
             });
+          };
+
+          /*
+           * Autarchy
+           */
+          let autarchy = consumptionData.map((value, index) => {
+            return (1 - (Utils.orElse(buyFromGridData[index], 0) / Utils.orElse(value, 0))) * 100;
+          }).map(value => {
+            if (value == null || isNaN(value)) {
+              return null;
+            }
+            else if (value >= 0) {
+              return value;
+            }
+            else {
+              return 0;
+            }
+          });
```

Comment:
Can be 'kind of' replaced with

              effectivePower = result.data['_sum/ProductionDcActualPower'].map((value, index) => {
                return Utils.subtractSafely(result.data['_sum/EssActivePower'][index], value);
              });